src/Razor/benchmarks/Microsoft.AspNetCore.Razor.Microbenchmarks/ProjectSystem/ProjectMutationBenchmark.cs

```diff
@@ -0,0 +1,101 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+#nullable disable
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using BenchmarkDotNet.Attributes;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+using Microsoft.CodeAnalysis.Razor.Workspaces;
+
+namespace Microsoft.AspNetCore.Razor.Microbenchmarks;
+
+public class ProjectMutationBenchmark : ProjectSnapshotManagerBenchmarkBase
+{
+    private readonly ProjectSnapshotManagerDispatcher _dispatcher;
+    public ProjectMutationBenchmark()
+        : base(100000)
+    {
+        _dispatcher = new SnapshotDispatcher(nameof(ProjectMutationBenchmark));
+    }
+
+    private Thread _addRemoveThread;
+    private Thread _readThread;
+
+    [IterationSetup]
+    public void Setup()
+    {
+        SnapshotManager = CreateProjectSnapshotManager(_dispatcher);
+    }
+
+    private DefaultProjectSnapshotManager SnapshotManager { get; set; }
+
+    [Benchmark(Description = "Does thread contention add/remove of documents", OperationsPerInvoke = 100)]
+    public async Task ProjectMutation_Mutates100kFiles()
+    {
+        await _dispatcher.RunOnDispatcherThreadAsync(() =>
+        {
+            SnapshotManager.ProjectAdded(HostProject);
+        }, CancellationToken.None);
```

Comment:
Should this happen in Setup?

Reply:
maybe... I wasn't sure if setup could be async.